### PR TITLE
[Refactor] Add require_version configuration in ti.init()

### PR DESCRIPTION
**Pull request by tczhangzhi**
Related issue = #3847

### CODE REVIEW

**Comment by tczhangzhi** [comment]
I suppose passing a tuple into `taichi.init` is not in line with the user's intuition to set required_version. Thus, I slightly modify the `required_version` API to permit a version string as input while refactoring `taichi.init`. A use case is as follows:
```
import taichi

taichi.init(require_version='0.8.0')
```
instead of:
```
import taichi

taichi.init(require_version=(0,8,0))
```
If everything is okay, I would finish a unit test and fix the broken usage of require_version in the taichi_elements project accordingly. Otherwise, I would like to improve the current API following other contributors' advice.

File: python/taichi/lang/__init__.py [hunk 1]
```diff
@@ -566,6 +568,39 @@ def init(arch=None,
                                                 daemon=True)
         check_version_thread.start()
 
+    # Check if installed version meets the requirements.
+    if require_version:
+        # Extract version number part (i.e. toss any revision / hash parts).
```
Comment (ailzhang) [inline python/taichi/lang/__init__.py:573]: let's still keep this in a separate function?

**Review by ailzhang — APPROVED** [review]
Thanks a lot for contributing! 
I really like the idea of using a string instead of tuple, another really nice to allow user only specifying `<major>.<minor>` as well. ;) 
Btw this file has been refactored to `python/taichi/lang/misc.py` in #4150 but resolving the merge conflict should be easy.

File: python/taichi/lang/__init__.py (hunk 1 above)
Comment (tczhangzhi) [reply python/taichi/lang/__init__.py:573]: okay

**Comment by tczhangzhi** [comment]
I recently finished further commits following your comments. Thanks for your review and happy new year! (*^▽^*)

**Comment by k-ye** [comment]
Thanks for the contribution! Would you resolve the merge conflicts when you have time, happy new year!

**Comment by tczhangzhi** [comment]
DONE. It is clear now (*^▽^*)

File: python/taichi/lang/misc.py [hunk 2]
```diff
@@ -179,11 +179,54 @@ def prepare_sandbox():
...
+        int(_ti_core.get_version_minor()),
+        int(_ti_core.get_version_patch()),
+    ]
+    # Match installed version and required version.
+    match = major == versions[0] and (
+        minor < versions[1] or minor == versions[1] and patch <= versions[2])
+
+    if not match:
+        print(f"Taichi version mismatch. required >= {major}.{minor}.{patch}")
+        print("Installed =", _ti_core.get_version_string())
```
Comment (ailzhang) [inline python/taichi/lang/misc.py:220]: nit: can we unify the two prints to one and put them directly in the Exception message below?

File: python/taichi/lang/misc.py [hunk 3]
```diff
@@ -208,6 +252,10 @@ def init(arch=None,
     # Check version for users every 7 days if not disabled by users.
     _version_check.start_version_check_thread()
 
+    # Check if installed version meets the requirements.
+    if not require_version is None:
```
Comment (ailzhang) [inline python/taichi/lang/misc.py:256]:
```suggestion
    if require_version is not None:
```

**Review by ailzhang — CHANGES_REQUESTED** [review]
Thanks! Only have a few nits left, would you mind also fix issues thrown by `pylint python/taichi/ --disable=all --enable=$(python scripts/generate_pylint_tags.py)` to pass the static checks?

**Comment by tczhangzhi** [comment]
Sorry for the late reply. Updated (ง •̀_•́)ง If there are any other comments, I will follow as soon as possible!

**Review by ailzhang — APPROVED** [review]
Thanks!

**Pull request merged**